lib/diagnostics_channel.js

```diff
@@ -0,0 +1,157 @@
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    try {
+      for (var i = 0; i < this._subscribers.length; i++) {
+        this._subscribers[i](data, this.name);
+      }
+    } catch (err) {
+      process.nextTick(() => {
+        triggerUncaughtException(err, false);
+      });
+    }
+  }
+}
+
+class Channel {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActiveChannel.prototype);
+    this._subscribers = [];
+    this.subscribe(subscription);
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = ObjectCreate(null);
+    this._subscribers = [];
+  }
+
+  channel(name) {
+    let channel;
+    const ref = this._channels[name];
+    if (ref) channel = ref.get();
+    if (channel) return channel;
+
+    if (typeof name !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('channel', ['string'], name);
+    }
+
+    channel = new Channel(name);
+
+    // New channels should check existing subscribers
+    for (let i = 0; i < this._subscribers.length; i++) {
+      const subscriber = this._subscribers[i];
+      channel.subscribe(subscriber);
+    }
+
+    this._channels[name] = new WeakReference(channel);
+    return channel;
+  }
+
+  subscribe(matcher, onMessage) {
+    // Namespace
+    if (typeof matcher === 'string') {
+      this.channel(matcher).subscribe(onMessage);
+      return;
+    }
+
+    let subscriber;
+    if (typeof matcher === 'function') {
+      subscriber = matcher;
+    } else if (typeof matcher === 'object' &&
+               typeof matcher.onMessage === 'function') {
+      subscriber = matcher;
+    }
+
+    if (typeof subscriber === 'undefined') {
+      throw new ERR_INVALID_ARG_TYPE(
+        'matcher', ['string', 'function', 'Subscriber'], matcher);
+    }
+
+    if (!ArrayPrototypeIncludes(this._subscribers, subscriber)) {
+      ArrayPrototypePush(this._subscribers, subscriber);
+
+      // New subscribers should check existing channels
+      for (const key in this._channels) {
+        const channel = this._channels[key].get();
+        if (channel) {
+          channel.subscribe(subscriber);
+        } else {
+          delete this._channels[key];
+        }
+      }
+    }
+  }
+
+  publish(name, data) {
+    return this.channel(name).publish(data);
+  }
+
+  shouldPublish(name) {
+    return this.channel(name).shouldPublish();
```

Comment:
Would you mind to add an if that checks if there is a channel before creating one?

Reply:
The `channel(...)` method already does that--it's a get-or-create method. Do you mean to _not_ create one in the `shouldPublish` method if it doesn't already exist? Because that seems reasonable.

Reply:
That's what I mean.